### ISSUE TITLE
Only move toward endstop if position is unknown.

### ISSUE THREAD

**Issue by whosawhatsis** [Smoothieware/Smoothieware#461]
Some machines have more physical failsafes than others. It would be good to have a configuration option (should probably be per-axis) to only allow an axis to move in the direction of homing (with endstop always enabled) whenever its position is not known. The position of an axis is not known when a machine first starts up and hasn't homed, or when the stepper is disabled allowing the motor to turn freely.

**Comment by sareph** [comment]
> so enable them as limit switches and you are protected :)

Yes, but this is problematic overkill. If limit switches are enabled, and You execute G1 X0 Y0 printer will go to 0x0, hit endstops and everything will stop until M999 or reset. For random god's sake, why? 0x0 is perfectly within limits. 
If You don't enable limit switches, you can happily execute G1 X-15 (what I did, because I copied end gcode from my other printer where this is allowable), smoothie will be very happy to smash into endstop, and break it. Aaaarghhh.
If endstop is triggered header under any circumstances should move toward that endstop, without halting everything else.

**Comment by pelrun** [comment]
Yeah, I was pretty flabbergasted when I found I had to disable the limit switches because going to X0 Y0 would kill everything.

**Comment by Zaaphod** [comment]
if going to X0 Y0 is killing everything, then you didn't set up or configure your machine correctly.  

Machines should be configured with limits both limits always active, then simply set alpha_min  -5 for example... now when you home the machine, X=-5  now do a G0 X0 Y0 and your X is 5mm away from touching the limit switch and returning to X0 at any speed will never trigger the limit switch.  
you don't need to even home off the min limit switch, you could home from the max switch and set alpha max to say 105,  now when you home your X=105 and moving to a maximum of X100 would never hit the switch.   Any good machine design would have the physical limits slightly outside the boundaries of the work table, so you would use alpha_min, alpha_max, beta_min, beta, max, gamma_min, and gamma_max as needed depending on your desired home direction to set your position after homing such that X0Y0Z0 is precisely in on the correct position on the machine table.   You don't move the limit switch to a position so that when it homes it's at 0... you move the switch so it's behind 0 so you can get back to 0 without hitting it.

I think the reason everyone is confused by this is they think that since they set a retract distance of say 5mm,  that it will set it to 0 after it does the retract, but it sets the axis before the retract happens, so if you have a retract of 5mm, then alpha_min should be -5 if you want it to be 0 after it gets done homing... but most machines would need to be something other than -5 because setting the position of a limit switch accurately is just a waste of time.  It's very difficult and not necessary.  Just put the switch well outside the work area, home the machine, now jog the machine until you get it to where you want 0 to be then set alpha_min to the negative distance you moved it, so if you moved it 13.412mm set alpha_min to -13.412 now home it again and go to 0 and you will be precisely at 0 and not hitting the switch.

**Comment by LucaBertuolo** [comment]
Thanks Zaaphod, very interesting...
I'm sill havin same issue on a corexy...

If i follow your council i will solve the problem when homing and got negative value on gcode...
But how can we solve the fact that if we have only min endstop, and X length of 300mm, and after homing i send a G0 X400 this will be very dangerous...
There is some way to set a max_alpla_lenght?

I've always use this parameter in other firmware like marlin or repetier...
With this limit, after homing the carriage will never go over max_alpla_lenght and this will no destroy my printer...

**Comment by wolfmanjm** [comment]
There is a PR #1149 that has soft endstops, if you want that feature give it a try.
